include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+        }
+
+        addr_info_iterator_t& operator++() WI_NOEXCEPT
+        {
+            this->operator+=(1);
+            return *this;
+        }
+
+        addr_info_iterator_t operator++(int) WI_NOEXCEPT
+        {
+            auto return_value = *this;
+            this->operator+=(1);
+            return return_value;
+        }
+
+        addr_info_iterator_t& operator+=(size_t offset) WI_NOEXCEPT
+        {
+            for (size_t count = 0; count < offset; ++count)
+            {
+                WI_ASSERT(m_addrinfo_ptr);
+                if (m_addrinfo_ptr)
+                {
+                    m_addrinfo_ptr = m_addrinfo_ptr->ai_next;
+
+                    if (m_addrinfo_ptr)
+                    {
+                        m_socket_address.set_sockaddr(m_addrinfo_ptr->ai_addr, m_addrinfo_ptr->ai_addrlen);
+                    }
+                    else
+                    {
+                        m_socket_address.reset();
+                    }
+                }
+            }
+
+            return *this;
+        }
+
+    private:
+        // non-ownership of this pointer - the parent class must outlive the iterator
+        T* m_addrinfo_ptr{nullptr};
+        ::wil::network::socket_address m_socket_address{};
+    }; // class addr_info_iterator_t
+
+    // begin() and end() enable range-based for loops after resolving a name
+    //
+    // For example:
+    //     wil::unique_addrinfo resolved_addr;
+    //     if (0 == GetAddrInfoW(L"name_to_resolve.xyz", nullptr, nullptr, resolved_addr.addressof()))
+    //     {
+    //         for (const auto& address : wil::network::addr_info_iterator{resolved_addr.get()})
+    //         {
+    //             // address == a wil::network::socket_address containing a resolved IP address
+    //         }
+    //     }
+
+    template <typename T>
+    addr_info_iterator_t<T> begin(addr_info_iterator_t<T> addrinfo) WI_NOEXCEPT
+    {
+        return addrinfo;
+    }
+
+    template <typename T>
+    constexpr addr_info_iterator_t<T> end(addr_info_iterator_t<T>) WI_NOEXCEPT
+    {
+        return {};
+    }
+
+    using addr_info_ansi_iterator = addr_info_iterator_t<ADDRINFOA>;
+    using addr_info_iterator = addr_info_iterator_t<ADDRINFOW>;
+    // not defining a type for ADDRINFOEXA as that type is formally __declspec(deprecated)
+    using addr_infoex_iterator = addr_info_iterator_t<ADDRINFOEXW>;
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+    // wil function to capture resolving IP addresses assigned to the local machine, throwing on error
+    // returning an RAII object containing the results
+    inline ::wil::unique_addrinfo resolve_local_addresses()
+    {
+        constexpr auto* local_address_name_string = L"";
+        ADDRINFOW* addrResult{};
+        if (::GetAddrInfoW(local_address_name_string, nullptr, nullptr, &addrResult) != 0)
+        {
+            THROW_WIN32(::WSAGetLastError());
+        }
+
+        return ::wil::unique_addrinfo{addrResult};

```

Comment:
Given the identical code below, is there any benefit to providing an exception-throwing wrapper of `GetAddrInfoW` that these two functions call? The benefit being both code reuse here and function availability to consumers